File: LLCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

//Handle the case of only one element differently
//Time Complexity: O(n).
//Space Complexity: O(1).
// The trick here is the mathematical property of distance. The distance 2(a+b) = a+b+c+b; imagine it as a circle and define the point where they meet and the respective distances. 

public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null) return null;
        
        ListNode slow = head;
        ListNode fast = head;
        boolean cycle = false;
        
        
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if(fast == slow){
                cycle = true;
                break;
            }
               
        }

        if(cycle == false) return null;
        
        if(cycle == true){
            slow = head;
            while(fast != slow){
                slow = slow.next;
                fast = fast.next;
             }
        }
  
        return slow;
    }
}

File: llcycle.py
'''
Time Complexity: O(n)
Space Complexity: O(1)
'''
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        slow = fast = head
        flag = 0
        while(fast and fast.next):
            slow = slow.next
            fast = fast.next.next
            if(slow==fast):
                flag = 1
                fast = head
                break
        if(flag == 1):
            while(slow != fast):
                slow = slow.next
                fast = fast.next
            return slow
        return None
        
